File: GanServer/src/com/xattacker/binary/OutputBinaryBuffer.java
package com.xattacker.binary;

import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.UnsupportedEncodingException;

public class OutputBinaryBuffer implements Closeable, BinaryWritable
{
	private OutputStream _out;
	
	public OutputBinaryBuffer(OutputStream aOut)
	{
		_out = aOut;
	}
	
	public void flush() throws IOException
	{
		if (_out != null)
		{
			_out.flush();
		}
	}

	public void close() throws IOException
	{
		if (_out != null)
		{
			_out.close();
		}
	}

	public void writeBinary(byte[] aData, int aOffset, int aLength) throws IOException
	{
		_out.write(aData, aOffset, aLength);
	}
	
	public void writeShort(short aShort) throws IOException
	{
		byte[] data = TypeConverter.shortToByte(aShort);

		writeBinary(data, 0, data.length);
	}
	
	public void writeInteger(int aInt) throws IOException
	{
		byte[] data = TypeConverter.intToByte(aInt);

		writeBinary(data, 0, data.length);
	}
	
	public void writeLong(long aLong) throws IOException
	{
		byte[] data = TypeConverter.longToByte(aLong);

		writeBinary(data, 0, data.length);
	}
	
	public void writeDouble(double aDouble) throws IOException
	{
		byte[] data = TypeConverter.doubleToByte(aDouble);

		writeBinary(data, 0, data.length);
	}
	
	public void writeString(String aStr) throws IOException, UnsupportedEncodingException
	{
		if (aStr == null || aStr.length() == 0)
		{
			writeInteger(0);
		}
		else
		{
			byte[] data = aStr.getBytes("UTF8");
			int length = data.length;

			writeInteger(length);
			writeBinary(data, 0, length);
		}
	}
	
	public void writeString(InputStream aIn, int aBufferSize) throws IOException
	{
		if (aIn == null || aIn.available() == 0)
		{
			writeInteger(0);
		}
		else
		{
			int length = aIn.available();
			writeInteger(length);

			byte temp[] = new byte[aBufferSize];
			int index = 0;
			
			while ((index = aIn.read(temp)) != -1)
			{
				writeBinary(temp, 0, index);
				flush();
			}
			
			temp = null;
		}
	}
	
	public void writeString(InputStream aIn) throws IOException
	{
		writeString(aIn, 5120);
	}
}

File: GanAndroidApp/GanClient/src/main/java/com/xattacker/binary/OutputBinaryBuffer.kt
package com.xattacker.binary

import java.io.Closeable
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

class OutputBinaryBuffer: Closeable, BinaryWritable<OutputBinaryBuffer>
{
    private var _out: OutputStream?

    constructor(outs: OutputStream)
    {
        _out = outs
    }

    @Throws(IOException::class)
    fun flush()
    {
        _out?.flush()
    }

    @Throws(IOException::class)
    override fun close()
    {
        _out?.close()
    }

    @Throws(IOException::class)
    override fun writeBinary(aData: ByteArray, aOffset: Int, aLength: Int): OutputBinaryBuffer
    {
        _out?.write(aData, aOffset, aLength)

        return this
    }

    @Throws(IOException::class)
    override fun writeShort(aShort: Short): OutputBinaryBuffer
    {
        val data = TypeConverter.shortToByte(aShort)
        writeBinary(data, 0, data.size)

        return this
    }

    @Throws(IOException::class)
    override fun writeInteger(aInt: Int): OutputBinaryBuffer
    {
        val data = TypeConverter.intToByte(aInt)
        writeBinary(data, 0, data.size)

        return this
    }

    @Throws(IOException::class)
    override fun writeLong(aLong: Long): OutputBinaryBuffer
    {
        val data = TypeConverter.longToByte(aLong)
        writeBinary(data, 0, data.size)

        return this
    }

    @Throws(IOException::class)
    override fun writeDouble(aDouble: Double): OutputBinaryBuffer
    {
        val data = TypeConverter.doubleToByte(aDouble)
        writeBinary(data, 0, data.size)

        return this
    }

    @Throws(IOException::class)
    override fun writeString(aStr: String): OutputBinaryBuffer
    {
        if (aStr.isEmpty())
        {
            writeInteger(0)
        }
        else
        {
            val data = aStr.toByteArray(Charsets.UTF_8)
            val length = data.size

            writeInteger(length)
            writeBinary(data, 0, length)
        }

        return this
    }

    @Throws(IOException::class)
    @JvmOverloads
    fun writeString(aIn: InputStream?, aBufferSize: Int = 5120): OutputBinaryBuffer
    {
        if (aIn == null || aIn.available() == 0)
        {
            writeInteger(0)
        }
        else
        {
            val length = aIn.available()
            writeInteger(length)

            val temp = ByteArray(aBufferSize)
            var index = 0

            while (true)
            {
                index = aIn.read(temp)
                if (index < 0)
                {
                    break
                }

                writeBinary(temp, 0, index)
                flush()
            }
        }

        return this
    }
}
